File: Data Structures & Algorithms/remove-node-from-end-of-linked-list/submission-6.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        dummy = ListNode(-1, head)
        p = dummy

        for _ in range(n):
            p = p.next

        prev = dummy
        while p.next:
            p = p.next
            prev = prev.next

        prev.next = prev.next.next

        return dummy.next

File: Data Structures & Algorithms/add-two-numbers/submission-6.js
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     constructor(val = 0, next = null) {
 *         this.val = val;
 *         this.next = next;
 *     }
 * }
 */

class Solution {
    /**
     * @param {ListNode} l1
     * @param {ListNode} l2
     * @return {ListNode}
     */
    addTwoNumbers(l1, l2) {
        const dummy = new ListNode()
        let curr = dummy
        let carry = 0

        while (l1 || l2 || carry) {
            const a1 = l1 ? l1.val : 0
            const a2 = l2 ? l2.val : 0
            
            const total = a1 + a2 + carry
            const newVal = total % 10 
            carry = Math.floor(total / 10)

            curr.next = new ListNode(newVal)
            curr = curr.next

            l1 = l1 ? l1.next : null
            l2 = l2 ? l2.next : null
        }

        return dummy.next
    }
}
